### ISSUE TITLE
Type forwarding seems to be ignored

### ISSUE THREAD

**Issue by SamuelDebruyn** [Azure/azure-functions-host#4527]
I am using the NuGet package `Google.Cloud.Firestore` which relies on some other packages where type forwarding is used. The forwarded types are not found and result in a `MissingMethodException`.

The issue occurs when debugging locally in VS Code on macOS. I haven't deployed this yet.

The same code works in a .NET Core console app.

A lot more details at the original issue: https://github.com/googleapis/google-cloud-dotnet/issues/3101

#### Investigative information

- Timestamp: Monday June 3th
- Function App version: 2.0

#### Repro steps

See original issue above

#### Expected behavior

Type forwarding is working and no `MissingMethodException`

#### Actual behavior

`MissingMethodException` when this calls into a forwarded type.

#### Known workarounds

?

#### Related information 

* Programming language used: C#
* Links to source: email me for this (sam@debruyn.dev)
* Bindings used: QueueTrigger

Csproj:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <AzureFunctionsVersion>v2</AzureFunctionsVersion>
    <RootNamespace>myappname</RootNamespace>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Azure.WebJobs.Extensions.Storage" Version="3.0.4"/>
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.24"/>
    <PackageReference Include="PuppeteerSharp" Version="1.17.1"/>
    <PackageReference Include="Google.Cloud.Firestore" Version="1.0.0-beta20"/>
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="google.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

**Comment by jskeet** [comment]
Note: the same issue occurs using the emulator from Windows and VS2019.

**Comment by jskeet** [comment]
I have a simpler repro now - there's no need for the Firestore dependency.

Project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <AzureFunctionsVersion>v2</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.Auth" Version="1.21.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.24" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```


Function source:

```csharp
using Google.Apis.Auth.OAuth2;
using Grpc.Auth;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;
using Microsoft.Extensions.Logging;
using System.Threading;
using System.Threading.Tasks;

namespace FunctionApp1
{
    public static class Function1
    {
        [FunctionName("Function1")]
        public static IActionResult Run(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req,
            ILogger log)
        {
            log.LogInformation("C# HTTP trigger function processed a request.");

            var tokenAccess = new DummyTokenAccess();
            // This succeeds
            tokenAccess.ToCallCredentials();
            // This fails
            tokenAccess.ToChannelCredentials();

            return new OkObjectResult("Passed");
        }
    }

    class DummyTokenAccess : ITokenAccess
    {
        public Task<string> GetAccessTokenForRequestAsync(string authUri = null, CancellationToken cancellationToken = default)
        {
            throw new System.NotImplementedException();
        }
    }
}
```

Exception:

```text
System.MissingMethodException
  HResult=0x80131513
  Message=Method not found: 'Grpc.Core.ChannelCredentials Grpc.Core.ChannelCredentials.Create(Grpc.Core.ChannelCredentials, Grpc.Core.CallCredentials)'.
  Source=Grpc.Auth
  StackTrace:
   at Grpc.Auth.GoogleGrpcCredentials.ToChannelCredentials(ITokenAccess googleCredential)
   at FunctionApp1.Function1.Run(HttpRequest req, ILogger log) in C:\Users\skeet\Test\Projects\Issues\google-cloud-dotnet\Issue3101\Issue3101\FunctionApp1\Function1.cs:line 26
```

This is very odd indeed:

- There are no different Grpc versions involved. We're compiling against Grpc.Auth 1.21.0 which was compiled against Grpc.Core 1.21.0, and that's the version present in both places.
- Looking at the IL for the method declaration and the call to it, from the DLLs in the bin directory, they look fine to me:

```il
// Declaration in Grpc.Core, in ChannelCredentials:
.method public hidebysig static class Grpc.Core.ChannelCredentials 
        Create(class Grpc.Core.ChannelCredentials channelCredentials,
               class [Grpc.Core.Api]Grpc.Core.CallCredentials callCredentials) cil managed


// Call in Grpc.Auth, in GoogleGrpcCredentials.ToChannelCredentials:
call       class [Grpc.Core]Grpc.Core.ChannelCredentials [Grpc.Core]Grpc.Core.ChannelCredentials::Create(class [Grpc.Core]Grpc.Core.ChannelCredentials,
                                                                                                                     class [Grpc.Core.Api]Grpc.Core.CallCredentials)
```

Note that the ToCallCredentials method works. CallCredentials is in Grpc.Core.Api; ChannelCredentials is in Grpc.Core.

I'm going to keep investigating this a bit to see if I can reproduce this without even a Grpc.Auth reference...

**Comment by jskeet** [comment]
Okay, even simpler repro - and an explanation.

First, the project file:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.2</TargetFramework>
    <AzureFunctionsVersion>v2</AzureFunctionsVersion>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="Grpc.Core" Version="1.21.0" />
    <PackageReference Include="Microsoft.NET.Sdk.Functions" Version="1.0.24" />
  </ItemGroup>
  <ItemGroup>
    <None Update="host.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
    </None>
    <None Update="local.settings.json">
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
      <CopyToPublishDirectory>Never</CopyToPublishDirectory>
    </None>
  </ItemGroup>
</Project>
```

Next, source code:

```cs
using Grpc.Core;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Azure.WebJobs;
using Microsoft.Azure.WebJobs.Extensions.Http;

namespace FunctionApp1
{
    public static class Function1
    {
        [FunctionName("Function1")]
        public static IActionResult Run(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest req)
        {
            CreateChannelCredentials(null);
            return new OkObjectResult("Passed");
        }

        private static ChannelCredentials CreateChannelCredentials(CallCredentials callCredentials) =>
            ChannelCredentials.Create(new SslCredentials(), callCredentials);
    }
}
```

This fails with the same MissingMethodException when you try to execute CreateChannelCredentials, when the JIT compiler works on that method - because it can't find the method.

The problem is that it's not loading Grpc.Core 1.21.0 at all - it's using 1.18.0 which is in `C:\Users\skeet\AppData\Local\AzureFunctionsTools\Releases\2.22.0\cli\`

If you update the `cli` directory to contain the relevant files from Grpc.Core and Grpc.Core.Api, and update `func.deps.json` accordingly, it all works - including Google.Cloud.Firestore.

So basically, this is a problem for any function needing to use a version of Grpc.Core later than the one embedded in the CLI application.

**Comment by AartBluestoke** [comment]
This is because the functions runtime in azure loads that specific version of the library , but the announced dependencies are often vague.  If you use features from a library newer that what is present things fail in the way described.
The same goes of you assume bugs are fixed. 
This happens for other libraries as well, e.g. newtonsoft.json https://github.com/Azure/azure-functions-host/issues/4049#issuecomment-505133197

**Comment by fabiocav** [comment]
Will investigate this as the version packaged with the CLI shouldn't cause this issue, so looks like there's a bug here. Do you happen to know it this is happening outside of the context of the CLI?

**Comment by jskeet** [comment]
@fabiocav: I don't know, I'm afraid. I assume that VS just launches the CLI? How can we test this outside the context of the CLI?

**Comment by fabiocav** [comment]
This was just in case you had tested this on Azure or in one of the container options. Wasn't really expecting that to be the case, but thought I'd ask. 

I'll assign this to the next sprint and take a closer look to see what is causing the issue.

**Comment by andreujuanc** [comment]
@fabiocav I just had this problem and when @jskeet wrote "func.deps.json" it opened my eyes and I just remembered this issue:
https://github.com/Azure/azure-functions-host/issues/4304

So I just added your fix for the WCF ref to my project file and boom, it worked:

 ``` xml
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <!--<Exec Command="copy $(OutDir)bin\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll $(OutDir)bin\System.Private.ServiceModel.dll" />-->
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="AfterPublish">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <!--<Exec Command="copy $(PublishDir)bin\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll $(PublishDir)bin\System.Private.ServiceModel.dll" />-->
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
```

**Comment by KasunKoswattha** [comment]
I'm trying to make this work on a mac. Is there a workaround or a solution ?

**Comment by jslaybaugh** [comment]
@KasunKoswattha On a mac, the `copy` command doesn't exist, so I've gotten past this issue by using 
```
<Target Name="PostBuild" AfterTargets="PostBuildEvent">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <!--<Exec Command="copy $(OutDir)bin\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll $(OutDir)bin\System.Private.ServiceModel.dll" />-->
    <Exec Command="cp $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostPublish" BeforeTargets="AfterPublish">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <!--<Exec Command="copy $(PublishDir)bin\runtimes\win\lib\netstandard2.0\System.Private.ServiceModel.dll $(PublishDir)bin\System.Private.ServiceModel.dll" />-->
    <Exec Command="cp $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
```

or in my specific case, where we need to run Windows AND OSX:
```
<Target Name="PostBuildWindows" AfterTargets="PostBuildEvent" Condition="'$(OS)' == 'Windows_NT'">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
    <Message Text="PostBuild Copied $(OutDir)bin\function.deps.json" />  
  </Target>
  <Target Name="PostPublishWindows" BeforeTargets="AfterPublish" Condition="'$(OS)' == 'Windows_NT'">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <Message Text="PostPublish Copied $(OutDir)bin\function.deps.json" />  
    <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
  </Target>
  <Target Name="PostBuildUnix" AfterTargets="PostBuildEvent" Condition="'$(OS)' != 'Windows_NT'">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <Exec Command="cp -r $(OutDir)$(ProjectName).deps.json $(OutDir)bin/function.deps.json" />
    <Message Text="PostBuild Copied $(OutDir)bin/function.deps.json" />  
  </Target>
  <Target Name="PostPublishUnix" BeforeTargets="AfterPublish" Condition="'$(OS)' != 'Windows_NT'">
    <!--https://github.com/Azure/azure-functions-host/issues/3568-->
    <Exec Command="cp -r $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin/function.deps.json" />
    <Message Text="PostPublish Copied $(OutDir)bin/function.deps.json" />  
  </Target>
```
However, I now have the following exception I can't seem to solve in my Azure function on OSX:

```
System.Private.CoreLib: Exception while executing function: CleanupQueue. Grpc.Core: The native method "grpcsharp_batch_context_recv_message_next_slice_peek" does not exist.
System.MissingMethodException: The native method "grpcsharp_batch_context_recv_message_next_slice_peek" does not exist
   at Grpc.Core.Internal.UnmanagedLibrary.GetNativeMethodDelegate[T](String methodName)
   at Grpc.Core.Internal.NativeMethods..ctor(UnmanagedLibrary library)
   at Grpc.Core.Internal.NativeExtension.LoadNativeMethods()
   at Grpc.Core.Internal.NativeExtension..ctor()
   at Grpc.Core.Internal.NativeExtension.Get()
   at Grpc.Core.GrpcEnvironment.GrpcNativeInit()
   at Grpc.Core.GrpcEnvironment..ctor()
   at Grpc.Core.GrpcEnvironment.AddRef()
   at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials, IEnumerable`1 options)
   at Google.Api.Gax.Grpc.ChannelPool.GetChannel(ServiceEndpoint endpoint, IEnumerable`1 channelOptions, ChannelCredentials credentials)
   at Google.Cloud.Firestore.FirestoreDb.Create(String projectId, FirestoreClient client)
   at Utilities.FirebaseUtility.GetEntryQueueStatus() in /Users/jslaybaugh/Projects/brushfire-netcore/src/Utilities/FirebaseUtility.cs:line 25
```

Can anyone help me figure out what this means? @jskeet ?

**Comment by jskeet** [comment]
@jslaybaugh: I'm afraid I can't easily tell what your context is, but I suspect it's the same underlying problem... I'm afraid I have no way of trying to support Azure Functions on a Mac.

**Comment by KasunKoswattha** [comment]
> @KasunKoswattha On a mac, the `copy` command doesn't exist, so I've gotten past this issue by using
> 
> ```
> <Target Name="PostBuild" AfterTargets="PostBuildEvent">
>     
>     
>     <Exec Command="cp $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
>   </Target>
>   <Target Name="PostPublish" BeforeTargets="AfterPublish">
>     
>     
>     <Exec Command="cp $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
>   </Target>
> ```
> 
> or in my specific case, where we need to run Windows AND OSX:
> 
> ```
> <Target Name="PostBuildWindows" AfterTargets="PostBuildEvent" Condition="'$(OS)' == 'Windows_NT'">
>     
>     <Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
>     <Message Text="PostBuild Copied $(OutDir)bin\function.deps.json" />  
>   </Target>
>   <Target Name="PostPublishWindows" BeforeTargets="AfterPublish" Condition="'$(OS)' == 'Windows_NT'">
>     
>     <Message Text="PostPublish Copied $(OutDir)bin\function.deps.json" />  
>     <Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
>   </Target>
>   <Target Name="PostBuildUnix" AfterTargets="PostBuildEvent" Condition="'$(OS)' != 'Windows_NT'">
>     
>     <Exec Command="cp -r $(OutDir)$(ProjectName).deps.json $(OutDir)bin/function.deps.json" />
>     <Message Text="PostBuild Copied $(OutDir)bin/function.deps.json" />  
>   </Target>
>   <Target Name="PostPublishUnix" BeforeTargets="AfterPublish" Condition="'$(OS)' != 'Windows_NT'">
>     
>     <Exec Command="cp -r $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin/function.deps.json" />
>     <Message Text="PostPublish Copied $(OutDir)bin/function.deps.json" />  
>   </Target>
> ```
> 
> However, I now have the following exception I can't seem to solve in my Azure function on OSX:
> 
> ```
> System.Private.CoreLib: Exception while executing function: CleanupQueue. Grpc.Core: The native method "grpcsharp_batch_context_recv_message_next_slice_peek" does not exist.
> System.MissingMethodException: The native method "grpcsharp_batch_context_recv_message_next_slice_peek" does not exist
>    at Grpc.Core.Internal.UnmanagedLibrary.GetNativeMethodDelegate[T](String methodName)
>    at Grpc.Core.Internal.NativeMethods..ctor(UnmanagedLibrary library)
>    at Grpc.Core.Internal.NativeExtension.LoadNativeMethods()
>    at Grpc.Core.Internal.NativeExtension..ctor()
>    at Grpc.Core.Internal.NativeExtension.Get()
>    at Grpc.Core.GrpcEnvironment.GrpcNativeInit()
>    at Grpc.Core.GrpcEnvironment..ctor()
>    at Grpc.Core.GrpcEnvironment.AddRef()
>    at Grpc.Core.Channel..ctor(String target, ChannelCredentials credentials, IEnumerable`1 options)
>    at Google.Api.Gax.Grpc.ChannelPool.GetChannel(ServiceEndpoint endpoint, IEnumerable`1 channelOptions, ChannelCredentials credentials)
>    at Google.Cloud.Firestore.FirestoreDb.Create(String projectId, FirestoreClient client)
>    at Utilities.FirebaseUtility.GetEntryQueueStatus() in /Users/jslaybaugh/Projects/brushfire-netcore/src/Utilities/FirebaseUtility.cs:line 25
> ```
> 
> Can anyone help me figure out what this means? @jskeet ?

Thanks  @jslaybaugh I will try this out. But we really need a solution for this. :(

**Comment by DJ4ddi** [comment]
I ran into the issue described on [StackOverflow](https://stackoverflow.com/questions/56708920/grpc-auth-could-not-load-type-grpc-core-callcredentials-from-assembly-grpc-c), leading me here.

Locally, both manually including a more recent package version of **Grpc.Core** via NuGet and the workaround posted by @andreujuanc worked. Unfortunately, neither did the trick for the remote end in the Azure cloud. Copying the dependency JSON changes nothing, while upgrading the package changes the original "Could not load type" error to
`Method not found: 'Grpc.Core.ChannelCredentials Grpc.Auth.GoogleGrpcCredentials.ToChannelCredentials(Google.Apis.Auth.OAuth2.ITokenAccess)'.`
So, while it does make a difference, I still can't run the function properly. Interestingly, also manually including **Grpc.Core.Api** restores the original error
`Could not load type 'Grpc.Core.CallCredentials' from assembly 'Grpc.Core.Api, Version=1.0.0.0, Culture=neutral, PublicKeyToken=d754f35622e28bad'.`
And in some combination of workarounds, I also managed to kill the library entirely, as indicated by the fact that `grpc_csharp_ext.x86.dll` failed to load.

Needless to say, I'm a bit confused. Why does basically every combination of proposed solutions work locally, but none of them solve the problem for the cloud runtime (which, according to some reports, shouldn't even be affected in the first place)?

**Comment by jskeet** [comment]
@DJ4ddi: If you've upgraded to Grpc.Core version 2.x, that would definitely cause problems for anything expecting 1.x, including (currently) all the Google Cloud client libraries. (These will soon be upgraded to Grpc.Core 2.x - at which point you won't be able to use them in the same app with anything 1.x-based, of course.)

**Comment by DJ4ddi** [comment]
The symptoms were the exact same when manually installing version 1.22 and 1.22.1.

**Comment by wikes82** [comment]
the workaround of copying *.deps.json work on local dev machine, but I'm still getting the error on publish to Azure Functions.

**Comment by jslaybaugh** [comment]
@wikes82 me too! But only if it runs through the publish on Azure DevOps. I've got a ticket open with them now about that. If I right click and tell it to do a zip deploy from VS (Windows) it works in Azure Functions for me with the following code:
```
	<Target Name="PostBuildWindows" AfterTargets="PostBuildEvent">
		<!-- https://github.com/Azure/azure-functions-host/issues/3568 -->
		<Exec Command="copy $(OutDir)$(ProjectName).deps.json $(OutDir)bin\function.deps.json" />
	</Target>
	<Target Name="PostPublishWindows" BeforeTargets="CreateZipFile">
		<!-- https://github.com/Azure/azure-functions-host/issues/3568 -->
		<Exec Command="copy $(PublishDir)$(ProjectName).deps.json $(PublishDir)bin\function.deps.json" />
	</Target>
```

Note the `BeforeTargets="CreateZipFile"` I'm now using.

**Comment by wikes82** [comment]
@jslaybaugh I can't find option to do zip deploy on my VS2019 Community

**Comment by wikes82** [comment]
I'm getting the following error now 
`Error loading native library. Not found in any of the possible locations: D:\home\site\wwwroot\bin\grpc_csharp_ext.x86.dll,D:\home\site\wwwroot\bin\runtimes/win/native\grpc_csharp_ext.x86.dll,D:\home\site\wwwroot\bin\../..\runtimes/win/native\grpc_csharp_ext.x86.dll`

I checked and the file grpc_csharp_ext.x86.dll is in /runtimes/win/native

**Comment by wikes82** [comment]
I fixed it by adding this line into @jslaybaugh BeforeTargets="CreateZipFile" 
` <Exec Command="copy $(PublishDir)runtimes\win\native\*.* $(PublishDir)bin\" />`

**Comment by KasunKoswattha** [comment]
Are we going to get a fix for this from Microsoft ?

**Comment by fabiocav** [comment]
This should be resolved. The fix was pushed a bit ago in version `1.0.30-beta2` of the SDK, but I just pushed an update with some enhancements. Can you please update your reference to `1.0.30-beta3` and retry?

**Comment by KasunKoswattha** [comment]
> This should be resolved. The fix was pushed a bit ago in version `1.0.30-beta2` of the SDK, but I just pushed an update with some enhancements. Can you please update your reference to `1.0.30-beta3` and retry?

Thanks @fabiocav . Let me try it out and get back to you

**Comment by KasunKoswattha** [comment]
@fabiocav I can confirm that I don't see the issue with 1.0.30-beta. I'm using it with .netcoreapp 2.1 on a mac. Thanks for helping on this. @jskeet Thank you for troubleshooting and providing information.

**Comment by Zastai** [comment]
Not related to the main issue, but is there any reason people are using <Exec> with copy/cp instead of <Copy> in their workarounds? Seems like a step backwards to start executing system utilities when MSBuild has built-in support.

**Comment by EthanSK** [comment]
This is still a problem...
